geopandas/geodataframe.py

```diff
@@ -1102,6 +1102,34 @@ def to_wkt(self, **kwargs):
 
         return df
 
+    def to_arrow(self, index=None, geometry_encoding="WKB", interleaved=True):
```

Comment:
Yes, I know, but the `interleaved` is consistent with an existing keyword in shapely. But maybe that's not a good enough reason ;)

Reply:
It's maybe slightly confusing to have both `interleaved` and `geometry_encoding` parameters. Could we have `geometry_encoding: "WKB" | "interleaved" | "separated"`?

Reply:
I was planning to allow to user to specify a more specific geoarrow type, like "point", "multipoints", etc (i.e. essentially anything that comes after `geoarrow.<>` in the extension type name). That would give users the possibility to control the exact output (eg they know they can have mix of polygons and multipolygons, and want to be sure to always export as multipolygon regardless of the actual values present)

If we do that, that wouldn't really match with allowing `"interleaved" | "separated"`

Reply:
GDAL has `GEOMETRY_ENCODING=[WKB/WKT/GEOARROW/GEOARROW_INTERLEAVED]`, which is in principle what @kylebarron is proposing. I tend to prefer that as well, rather than having an additional keyword (which applies only if you change the default value of `geometry_encoding`).

> I was planning to allow to user to specify a more specific geoarrow type, like "point", "multipoints", etc

I suppose we should rather have a method that coerces geometries in a series into Multi* types, so you can do that explicitly before calling `to_arrow`. I suppose that was the original idea behind [`collect`](https://github.com/geopandas/geopandas/blob/1f5789f32010428d760eae070ae741e63f35d7b1/geopandas/tools/util.py#L13).

Reply:
Yes, there is https://github.com/shapely/shapely/issues/1617 as a feature request for this in shapely.



> I suppose we should rather have a method that coerces geometries in a series into Multi* types, so you can do that explicitly before calling `to_arrow`.

That is slightly less efficient, because there is no need to create new geometries in an intermediate step, since `to_ragged_array` can do this on the fly.

Reply:
Also, assume we would add a `geoarrow.collection` type using some kind of union type (something we have been discussing). At that point you want to be able to specify this as the encoding, independently from interleaved vs separated.

(I certainly agree it is not great to have that separate keyword which then is ignored for the default of WKB)

Reply:
Okay then. I am fine with either. 

The default will eventually change from WKB anyway, no?

Reply:
> Ah, I didn't know that GDAL stuck the coordinate type with the encoding. It would mostly just be nice if constructing a GeoArrow type was as consistent as possible everywhere.

Note that in the case of GDAL, this is for writing GeoParquet, which officially only supports the separated structs. That's the reason they made "GEOARROW" mean that, and only added "GEOARROW_INTERLEAVED" because originally "GEOARROW" did that.

For example, for our `to_parquet` specifying geoarrow as the encoding will also always give you the separated coords out of the box (I wasn't planning to allow writing interleaved coords).

> If coordinate type is part of the encoding, `encoding`/`geometry_type`/`dimensions` is sufficient to parameterize all of the current types. (This makes it more verbose to do something like add 64-bit list offset support, which might be an encoding of LARGE_GEOARROW, or add float coordinate support, but those might never happen).

Yeah, that's one of the reasons I am hesitant to stuff everything in the encoding name. Because for GeoArrow itself, in theory we don't need a new extension name, we could just say that eg a "geoarrow.polygon" can use both list or large list under the hood (just as we already allow struct or fixed size list), and in the future potentially list view etc. 
I think in a geopandas `to_arrow` interface I would rather have separate keywords to control those options?

Reply:
I'm fine with separate options here.

Reply:
Yes, and also later on if we want to improve this interface by using names for the encoding or whathever, it should always be relatively straightforward to keep supporting the interface we come up now for back compat. Getting _something_ out there for now is more important ;)

Reply:
> Maybe `def to_arrow(self, index=None, *, geometry_encoding=None, interleaved=None)` would leave some flexibility to sort out the details in a backwards-compatible way?

@paleolimbot with ` geometry_encoding=None` being None by default, you meant to have the user required to always specify some encoding (i.e. not actually default to WKB) ?